R/scale-discrete-.r

```diff
@@ -141,35 +141,104 @@ ScaleDiscretePosition <- ggproto("ScaleDiscretePosition", ScaleDiscrete,
   }
 )
 
-# TODO: This is a clear candidate for vctrs once we adopt it
-new_mapped_discrete <- function(x) {
-  if (is.null(x)) {
-    return(x)
-  }
-  if (!is.numeric(x)) {
-    cli::cli_abort("{.cls mapped_discrete} objects can only be created from numeric vectors")
-  }
-  class(x) <- c("mapped_discrete", "numeric")
-  x
+new_mapped_discrete <- function(x = double()) {
+  vec_assert(x, double())
+  obj <- new_vctr(x, class = "ggplot2_mapped_discrete")
+  # vctrs does not support inheriting from numeric base class
+  class(obj) <- c(class(obj), "numeric")
+  obj
+}
+mapped_discrete <- function(x = double()) {
+  if (is.null(x)) return(NULL)
+  if (is.array(x)) x <- as.vector(x)
+  new_mapped_discrete(vec_cast(x, double()))
+}
+is_mapped_discrete <- function(x) inherits(x, "ggplot2_mapped_discrete")
+#' @export
+format.ggplot2_mapped_discrete <- function(x, ...) format(vec_data(x), ...)
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.ggplot2_mapped_discrete <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.numeric <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.numeric.ggplot2_mapped_discrete <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.double <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.double.ggplot2_mapped_discrete <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.integer <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.integer.ggplot2_mapped_discrete <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.character <- function(x, y, ...) character()
+#' @export
+vec_ptype2.character.ggplot2_mapped_discrete <- function(x, y, ...) character()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.factor <- function(x, y, ...) new_mapped_discrete()
```

Comment:
Our type hierarchy is like:

```
character
   ^
   |
   |
factor
```

So it felt weird to me that:

```
<chr> + <mapped_discrete> = <chr>

# but

<fct> + <mapped_discrete> = <mapped_discrete>
```

In my head that only works if you consider a mapped_discrete as something between a character and a factor, is that right?

Side question - I'd just like to confirm that you really do want these ptype2 methods related to factors and characters, i.e. are you sure you want this to work?

```
vec_c("x", mapped_discrete(1))
```

If mapped_discrete is a numeric value, I don't feel like it should be combine-able with a character vector. `vec_c("x", 1)` fails in vctrs.

Reply:
see my main comment.